### PR TITLE
Kotlin: extract annotations

### PR DESCRIPTION
**Pull request by smowton**
This establishes annotation extraction on a par with the Java extractor (annotations for classes/interfaces, value parameters, fields, callables, and enum entries, but noteworthily not for type parameters or type usages, which require database schema work.) Not annotating type usages particularly means that we don't yet need to worry about nullability annotations.

Wrinkles:
* Certain meta-annotations require translation between Kotlin and JVM universes; the code for this is largely lifted with attribution from the Kotlin compiler.
* Repeatable annotations are particularly difficult, requiring a synthetic container class when specified in Kotlin.
* All annotation expressions require full global identifiers, because they can be extracted when seen from outside as an external dependency.

### CODE REVIEW

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 1]
```diff
@@ -4101,6 +4451,163 @@ open class KotlinFileExtractor(
         extractExpressionExpr(loop.condition, callable, id, 0, id)
     }
 
+    @Suppress("UNCHECKED_CAST")
+    private fun <T : DbExpr> exprIdOrFresh(id: Label<out DbExpr>?) = id as? Label<T> ?: tw.getFreshIdLabel()
```
Comment (igfoo) [inline java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:4455]: I think we can avoid the warning suppression here by using something like `id?.cast() ?: tw.getFreshIdLabel()`

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 2]
```diff
@@ -4101,6 +4451,163 @@ open class KotlinFileExtractor(
...
+                }
+            }
+            v is String -> {
+                exprIdOrFresh<DbStringliteral>(overrideId).also { id ->
+                    val type = useType(e.type)
+                    val locId = tw.getLocation(e)
+                    tw.writeExprs_stringliteral(id, type.javaResult.id, parent, idx)
+                    tw.writeExprsKotlinType(id, type.kotlinResult.id)
+                    extractExprContext(id, locId, enclosingCallable, enclosingStmt)
+                    tw.writeNamestrings(toQuotedLiteral(v.toString()), v.toString(), id)
```
Comment (tamasvajk) [inline java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:4597]: This change results in a lot of changes to expected files. Would it make sense to do this in a separate PR to make the diff cleaner?

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 3]
```diff
@@ -450,7 +464,406 @@ open class KotlinFileExtractor(
         extractDeclInitializers(c.declarations, false) { Pair(blockId, obinitId) }
     }
 
-    val jvmStaticFqName = FqName("kotlin.jvm.JvmStatic")
+    // Adapted from RepeatedAnnotationLowering.kt
```
Comment (tamasvajk) [inline java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:467]: This file is already quite large, can we easily move these adapted functions into a separate file?

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 4]
```diff
@@ -450,7 +464,406 @@ open class KotlinFileExtractor(
...
+        containerClass: IrClass,
+        entries: List<IrConstructorCall>,
+    ): IrConstructorCall? {
+        val annotationType = annotationClass.typeWith()
+        val containerConstructor = containerClass.primaryConstructor
+        if (containerConstructor == null) {
+            logger.warnElement("Expected container class to have a primary constructor", containerClass)
+            return null
+        } else {
+            return IrConstructorCallImpl.fromSymbolOwner(containerClass.defaultType, containerConstructor.symbol).apply {
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt]: All other comments are addressed, but this one is considerable work re: either manually walking through the process or customising the class-extraction routines to accept various synthetic class attributes. I appreciate I've also shied away from IR generation before, but this one finally crossed over the degree of difficulty to make me do it and try to convince myself there wouldn't be untoward side-effects instead.

**Comment by smowton** [comment]
@tamasvajk @igfoo all dependent PRs have landed; this is ready for review

**Review by tamasvajk — COMMENTED** [review]
Looks plausible to me, it's a low confidence review. (The PR is rather large) There are some todo comments in this PR, should the list of open tasks be added to the internal issue regarding annotation extraction?

File: java/ql/consistency-queries/cfgDeadEnds.ql [hunk 5]
```diff
@@ -53,6 +53,8 @@ predicate shouldBeDeadEnd(ControlFlowNode n) {
   not exists(n.getFile().getRelativePath()) // TODO
   or
   n = any(ConstCase c).getValue(_) // TODO
+  or
+  n instanceof ErrorExpr // TODO
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/ql/consistency-queries/cfgDeadEnds.ql:57]: This is from your draft of annotation extraction, but we can try removing it and see what happens

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 6]
```diff
@@ -16,24 +14,52 @@ import org.jetbrains.kotlin.descriptors.java.JavaVisibilities
...
+import org.jetbrains.kotlin.ir.util.isSuspend
+import org.jetbrains.kotlin.ir.util.isSuspendFunctionOrKFunction
+import org.jetbrains.kotlin.ir.util.isVararg
+import org.jetbrains.kotlin.ir.util.kotlinFqName
+import org.jetbrains.kotlin.ir.util.packageFqName
+import org.jetbrains.kotlin.ir.util.parentAsClass
+import org.jetbrains.kotlin.ir.util.parentClassOrNull
+import org.jetbrains.kotlin.ir.util.primaryConstructor
+import org.jetbrains.kotlin.ir.util.render
+import org.jetbrains.kotlin.ir.util.target
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt]: It turned out there was a name clash between something imported elsewhere and a member of `ir.util` for certain versions of Kotlin (I forget which symbol was at fault)

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 7]
```diff
@@ -187,7 +217,7 @@ open class KotlinFileExtractor(
                 is IrField -> {
                     val parentId = useDeclarationParent(getFieldParent(declaration), false)?.cast<DbReftype>()
                     if (parentId != null) {
-                        extractField(declaration, parentId)
+                        extractField(declaration, parentId, extractFunctionBodies)
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:220]: I believe this is right; have added a comment to clarify.

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 8]
```diff
@@ -457,7 +489,140 @@ open class KotlinFileExtractor(
...
+            is IrConst<*> -> {
+                extractConstant(v, parent, idx, null, null, overrideId = exprId())
+            }
+            is IrGetEnumValue -> {
+                extractEnumValue(v, parent, idx, null, null, extractTypeAccess = extractEnumTypeAccesses, overrideId = exprId())
+            }
+            is IrClassReference -> {
+                val classRefId = exprId()
+                val typeAccessId = tw.getLabelFor<DbUnannotatedtypeaccess>("@\"annotationExpr;{$classRefId};0\"")
+                extractClassReference(v, parent, idx, null, null, overrideId = classRefId, typeAccessOverrideId = typeAccessId, useJavaLangClassType = true)
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:587]: `useJavaLangClassType = true` is doing that for us

File: java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt [hunk 9]
```diff
@@ -457,7 +489,140 @@ open class KotlinFileExtractor(
...
+                // Note we lose any arguments to @java.lang.Deprecated that were written in source.
+                IrConstructorCallImpl.fromSymbolOwner(
+                    UNDEFINED_OFFSET, UNDEFINED_OFFSET, jldConstructor.returnType, jldConstructor.symbol, 0
+                )
+    }
+
+    private fun extractAnnotations(c: IrAnnotationContainer, annotations: List<IrConstructorCall>, parent: Label<out DbExprparent>, extractEnumTypeAccesses: Boolean) {
+        val origin = when(c) {
+            is IrDeclaration -> c.origin
+            else -> null
```
[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:513]: At present I believe nothing. In future, `IrFile` and `IrType` would be possible, but we don't extract those kinds of annotations yet.

[the remark this answers is not in the document]
Comment (smowton) [reply java/kotlin-extractor/src/main/kotlin/KotlinFileExtractor.kt:513]: I'll log when this happens though.

File: java/ql/consistency-queries/cfgDeadEnds.ql (hunk 5 above)
[the remark this answers is not in the document]
Comment (tamasvajk) [reply java/ql/consistency-queries/cfgDeadEnds.ql:57]: I don't remember why it was added. If you don't have the bandwidth now to check it, then leave it as it is. I'll check later, and probably add some comments why this is added.

**Review by tamasvajk — APPROVED** [review]
